### DISCUSSION TITLE
Vertical coupler optimisation for 2 polarisation states of an incident source i.e. Ex and Ey

### DISCUSSION THREAD

**Discussion by jsco7509** [NanoComp/meep#2936] [General]
Hi gang,

How do I perform topology optimisation of a vertical coupler attempting to maximise the coupling efficiency for an incident plane wave with two orthogonal polarisation states? i.e. Source_1 = Ex and Source_2 = Ey. (For 1 frequency only )

Thanks

**Comment by jsco7509** [comment]
When optimisation is performed for different frequencies it seems like this can be simply implemented using the "frequencies" Args in:

opt = mpa.OptimizationProblem(
    simulation=sim,
    objective_functions=[J1],
    objective_arguments=ob_list,
    design_regions=[design_region],
    **frequencies=frequencies,**
    maximum_run_time=2000,
)

However this is not possible for the different polarisation states. Therefore, I believe this would need to be done by defining sources with difference components i.e. 

sourceEx = [mp.Source(mp.GaussianSource(fcen, fwidth=width),
                     component=mp.Ex,
                     center=mp.Vector3(0, 0, -2),
                     size=mp.Vector3(3, 3, 0)),
           ]
sourceEy = [mp.Source(mp.GaussianSource(fcen, fwidth=width),
                     component=mp.Ey,
                     center=mp.Vector3(0, 0, -2),
                     size=mp.Vector3(3, 3, 0)),
    ]

Then I'm not sure how to update mp.Simulation and the **objective list** and **function**.

**Reply by smartalecH** [reply]
The objective function will assume the objective arguments are an array with the same shape as the `frequencies` parameter in your optimization configuration. You might be confusing what's going on here:

The sources are prescribed in the _time domain_ and hence excite a continuum of frequencies in your simulation. So you can effectively simulate _all_ frequencies (within your prescribed band) using those two polarized sources. 

To do any processing on these fields in the frequency domain, however, you have to use a DTFT at a series of specified frequencies (meep does all this under the hood). And the optimization object and objective function syntax follows this logic. 

The output of your objective function _can still be a vector_, and meep with return a full Jacobian here. But your optimizer may not know how to handle that... so we typically perform some kind of _reduction_ in the objective function (eg a mean) and meep will properly backpropagate through that.

**Reply by jsco7509** [reply]
Thank you @smartalecH

**Comment by jsco7509** [comment]
In regards to performing optimisation for a coupler using an incident plane wave source with **2 different polarisation states** I set up my simulation in the following way:

```
sourceEy = [mp.Source(mp.GaussianSource(fcen, fwidth=width),
                     component=mp.Ey,
                     center = source_center,
                     size = source_size),]
sourceEx = [mp.Source(mp.GaussianSource(fcen, fwidth=width),
                     component=mp.Ex,
                     center = source_center,
                     size=source_size)]

simEy = mp.Simulation(
    cell_size=cell_size,
    boundary_layers=pml_layers,
    geometry=geometry,
    sources=sourceEy,
    dimensions=3,
    resolution=resolution,
)

simEx = mp.Simulation(
    cell_size=cell_size,
    boundary_layers=pml_layers,
    geometry=geometry,
    sources=sourceEx,
    dimensions=3,
    resolution=resolution,
)

mode = 1

TEEy = mpa.EigenmodeCoefficient(
    simEy,
    mp.Volume(
        center=mp.Vector3(x=-4),
        size=mp.Vector3(0, 3, 0.3),
    ),
    mode,
)

TEEx = mpa.EigenmodeCoefficient(
    simEx,
    mp.Volume(
        center=mp.Vector3(x=-4),
        size=mp.Vector3(0, 3, 0.3),
    ),
    mode,
)


ob_list = [TEEy, TEEx]

def J1(TEEy, TEEx):
    power = npa.abs(TEEy) ** 2 + npa.abs(TEEx) ** 2
    return power

opt = mpa.OptimizationProblem(
    simulation=simEy,
    objective_functions=[J1],
    objective_arguments=ob_list,
    design_regions=[design_region],
    frequencies=frequencies,
    maximum_run_time=2000,
)

```

**Reply by jsco7509** [reply]
This, to me, seems like the more 'pure' way of optimising for 2 different polarisation states (Ex & Ey) of an incident plane wave. However I understand that it is more computationally expensive. (NB Here I want to couple both states into a single waveguide, similar to https://doi.org/10.1364/OL.36.000796)

I note however, that you seemed to construct your simulation in the reciprocal manner. Launching a source in the waveguide, then maximising the outcoupled efficiency using Near-to-Far-Field. Obviously you get a very nice result....

is the approach I use valid? I am also concerned about only calling 1 sim (simulation=simEy) in the optimisation problem yet  using simEx and simEy in the mpa.EigenmodeCoefficient:


```

TEEy = mpa.EigenmodeCoefficient(
    simEy,
    mp.Volume(
        center=mp.Vector3(x=-4),
        size=mp.Vector3(0, 3, 0.3),
    ),
    mode,
)

TEEx = mpa.EigenmodeCoefficient(
    simEx,
    mp.Volume(
        center=mp.Vector3(x=-4),
        size=mp.Vector3(0, 3, 0.3),
    ),
    mode,
)

opt = mpa.OptimizationProblem(
    simulation=simEy,
    objective_functions=[J1],
    objective_arguments=ob_list,
    design_regions=[design_region],
    frequencies=frequencies,
    maximum_run_time=2000,
)
```
